examples/vision/focal_modulation_network.py

```diff
@@ -0,0 +1,1220 @@
+Author: [Aritra Roy Gosthipaty](https://twitter.com/ariG23498), [Ritwik Raha](https://twitter.com/ritwik_raha)
+Date created: 2023/01/25
+Last modified: 2023/01/25
+Description: Image classification with Focal Modulation Networks.
+"""
+"""
+# Introduction
+
+This tutorial aims to provide a comprehensive guide to the implementation of
+[Focal Modulation Networks](https://arxiv.org/abs/2203.11926), as
+presented in the academic paper by Yang et. al. 
+
+This tutorial will provide a formal, minimalistic approach to implementing Focal
+Modulation Networks and explore its potential applications in the field of Deep Learning.
+
+## The Problem Statement
+
+The architecture of Transformer ([Vaswani et. al](https://arxiv.org/abs/1706.03762)),
+which has become the de facto standard in most Natural Language Processing tasks, has
+also been applied to the field of computer vision through the seminal work of Vision
+Transformers ([Dosovitskiy et. al](https://arxiv.org/abs/2010.11929v2)).
+
+> In Transformers, the self-attention (SA) is arguably the key to its success which
+enables input-dependent global interactions, in contrast to convolution operation which
+constraints interactions in a local region with a shared kernel.
+
+The **attention** module is mathematically written as
+
+$$\text{Attention} = \text{softmax}(\frac{QK^{T}}{\sqrt{d_k}})V$$
+
+where:
+- $Q$ is the query
+- $K$ is the key
+- $V$ is the value
+- $d_k$ is the dimension of the key
+
+With **self-attention** the query, key, and value are all sourced from the input
+sequence. Let's rewrite the attention equation for self-attention.
+
+$$\text{Self Attention} = \text{softmax}(\frac{QQ^{T}}{\sqrt{d_q}})Q$$
+
+Upon looking at the equation of self-attention we see that it is a quadratic equation.
+This means that as number of tokens increase, so does the computation time (cost too). To
+mitigate this problem, and also to make the Transformer more interpretable Yang et. al
+have tried to replace the Self Attention module with better componenets.
+
+## The Solution
+
+Yang et. al introduce the revolutionary Focal Modulation _Layer_, poised to serve as a
+seamless replacement for the Self Attention Layer. The layer boasts high
+interpretability, making it a valuable tool for Deep Learning practitioners.
+
+In this tutorial, we will delve into the practical application of this layer by training
+the entire model on the CIFAR-10 dataset and visually interpreting the layer's
+performance.
+
+Note: We have tried aligning our implementation with the
+[official implementation](https://github.com/microsoft/FocalNet) as
+much as we could.
+"""
+
+"""
+# Setup and Imports
+
+`tensorflow-addons` is required as we will use the AdamW optimizer to train our model.
+"""
+
+"""shell
+!pip install --upgrade --quiet tensorflow-addons
+"""
+
+import numpy as np
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+from tensorflow_addons.optimizers import AdamW
```

Comment:
Please use the one from Keras -- `keras.optimizers.experimental.AdamW` (or  just`keras.optimizers.AdamW` in 2.12)